From a4871ee3fba40b700b6722ce9dfb582198ad3628 Mon Sep 17 00:00:00 2001
From: Vitaly Prosyak <vitaly.prosyak@amd.com>
Date: Fri, 9 Sep 2016 17:20:27 -0400
Subject: [PATCH 0917/1722] drm/amd/dal: Hot-plug fails to light up on MST
 displays

 Add check for bit MSA_TIMING_PAR_IGNORED from dpcd 7.
 Allow freesync when MSA_TIMING_PAR_IGNORED equal to 1.

Change-Id: I12b4f5aa885a5b3cf541fac4c726e400e7580147
Signed-off-by: Vitaly Prosyak <vitaly.prosyak@amd.com>
Acked-by: Harry Wentland <harry.wentland@amd.com>
Signed-off-by: Kalyan Alle <kalyan.alle@amd.com>
---
 .../gpu/drm/amd/dal/amdgpu_dm/amdgpu_dm_types.c    | 29 ++++++++++++++++++++--
 1 file changed, 27 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/amd/dal/amdgpu_dm/amdgpu_dm_types.c b/drivers/gpu/drm/amd/dal/amdgpu_dm/amdgpu_dm_types.c
index 0df5e7e..369e337 100644
--- a/drivers/gpu/drm/amd/dal/amdgpu_dm/amdgpu_dm_types.c
+++ b/drivers/gpu/drm/amd/dal/amdgpu_dm/amdgpu_dm_types.c
@@ -3092,12 +3092,27 @@ int amdgpu_dm_atomic_check(struct drm_device *dev,
 	return ret;
 }
 
+static bool is_dp_capable_without_timing_msa(
+		struct dc *dc,
+		struct amdgpu_connector *amdgpu_connector)
+{
+	uint8_t dpcd_data;
+	bool capable = false;
+	if (amdgpu_connector->dc_link &&
+	    dc_read_dpcd(dc, amdgpu_connector->dc_link->link_index,
+			 DP_DOWN_STREAM_PORT_COUNT,
+			 &dpcd_data, sizeof(dpcd_data)) )
+		capable = dpcd_data & DP_MSA_TIMING_PAR_IGNORED? true:false;
+
+	return capable;
+}
 void amdgpu_dm_add_sink_to_freesync_module(
 		struct drm_connector *connector,
 		struct edid *edid)
 {
 	int i;
 	uint64_t val_capable;
+	bool edid_check_required;
 	struct detailed_timing *timing;
 	struct detailed_non_pixel *data;
 	struct detailed_data_monitor_range *range;
@@ -3107,15 +3122,25 @@ void amdgpu_dm_add_sink_to_freesync_module(
 
 	struct drm_device *dev = connector->dev;
 	struct amdgpu_device *adev = dev->dev_private;
-
+	edid_check_required = false;
 	if (!amdgpu_connector->dc_sink) {
 		DRM_ERROR("dc_sink NULL, could not add free_sync module.\n");
 		return;
 	}
 	if (!adev->dm.freesync_module)
 		return;
+	/*
+	 * restrict for now freesync only for dp and edp
+	 */
+	if (amdgpu_connector->dc_sink->sink_signal == SIGNAL_TYPE_DISPLAY_PORT
+	    || amdgpu_connector->dc_sink->sink_signal == SIGNAL_TYPE_EDP) {
+		edid_check_required = is_dp_capable_without_timing_msa(
+					adev->dm.dc,
+					amdgpu_connector);
+	}
 	val_capable = 0;
-	if (edid->version > 1 || (edid->version == 1 && edid->revision > 1)) {
+	if (edid_check_required == true && (edid->version > 1 ||
+	   (edid->version == 1 && edid->revision > 1))) {
 		for (i = 0; i < 4; i++) {
 
 			timing	= &edid->detailed_timings[i];
-- 
2.7.4

